elasticsearch-transport/lib/elasticsearch/transport/transport/http/manticore/pool.rb

```diff
@@ -0,0 +1,201 @@
+module Elasticsearch
+  module Transport
+    module Transport
+      module HTTP
+        class Manticore
+          class Pool
+            attr_reader :logger
+
+            def initialize(logger, adapter, healthcheck_path="/", urls=[], resurrect_interval=5)
+              @logger = logger
+              @state_mutex = Mutex.new
+              @url_info = {}
+              @adapter = adapter
+              @healthcheck_path = healthcheck_path
+              @stopping = false
+              @resurrect_interval = resurrect_interval
+              @resurrectionist = start_resurrectionist
+              update_urls(urls)
+            end
+
+            def close
+              logger.info  "Stopping resurrectionist" if logger
+              stop_resurrectionist
+              logger.info  "Waiting for open manticore connections" if logger
+              wait_for_open_connections
+              logger.info("Closing adapter #{@adapter}") if logger
+              @adapter.close
+            end
+
+            def wait_for_open_connections
+              until open_connections.empty?
+                logger.info "Blocked on shutdown to to open connections #{@state_mutex.synchronize {@url_info}}" if logger
+                sleep 1
+              end
+            end
+
+            def open_connections
+              @state_mutex.synchronize { @url_info.values.select {|v| v[:open] > 0 } }
+            end
+
+            def alive_urls_count
+              @state_mutex.synchronize { @url_info.values.select {|v| !v[:dead] }.count }
+            end
+
+            def start_resurrectionist
+              Thread.new do
+                last_resurrect = Time.now
+                until @state_mutex.synchronize { @stopping } do
+                  if Time.now-last_resurrect >= @resurrect_interval
+                    last_resurrect = Time.now
+                    resurrect_dead!
+                  end
+                end
+              end
+            end
+
+            def resurrect_dead!
+              # Try to keep locking granularity low such that we don't affect IO...
+              @state_mutex.synchronize { @url_info.select {|url,meta| meta[:dead] } }.each do |url,meta|
+                begin
+                  @logger.info("Checking url #{url} with path #{@healthcheck_path} to see if node resurrected! ")
+                  perform_request_to_url(url, "GET", @healthcheck_path)
+                  # If no exception was raised it must have succeeded!
+                  logger.warn("Resurrected connection to dead ES instance at #{url}")
+                  @state_mutex.synchronize { meta[:dead] = false }
```

Comment:
Any value that may be accessed from another thread MUST by synchronized per the JMM. 

I'm sure that the Hash object in question is not threadsafe so that's definitely not a safe idea to do otherwise.